### PR TITLE
Feat/sms support media and binary messsages

### PR DESCRIPTION
**Pull request by Dovchik**
Add support for distinct types of Media, Binary and Text SMS for Batches

### CODE REVIEW

File: src/Sinch/SMS/Batches/Send/BatchBase.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        ///     <see href="https://community.sinch.com/t5/SMS/How-do-I-assign-a-callback-URL-to-an-SMS-service-plan/ta-p/8414">here</see>
+        ///     .
+        /// </summary>
+        public Uri CallbackUrl { get; set; }
+
+        /// <summary>
+        ///     The client identifier of a batch message.
+        ///     If set, the identifier will be added in the delivery report/callback of this batch
+        /// </summary>
+        public string ClientReference { get; set; }
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/BatchBase.cs:114]: Not tagged as required from specs

File: src/Sinch/SMS/Batches/Send/BatchBase.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        /// <summary>
+        ///     If set to <c>true</c>, then
+        ///     <see
+        ///         href="https://developers.sinch.com/docs/sms/api-reference/sms/tag/Batches/#tag/Batches/operation/deliveryFeedback">
+        ///         feedback
+        ///     </see>
+        ///     is expected after successful delivery.
+        /// </summary>
+        public bool FeedbackEnabled { get; set; } = false;
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/BatchBase.cs:124]: Not tagged as required from specs

File: src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        public override SmsType Type { get; } = SmsType.MtMedia;
+
+        /// <summary>
+        ///     Default: <c>false</c>.<br/><br/>
+        ///     Whether or not you want the media included in your message to be checked against
+        ///     <see href="https://developers.sinch.com/docs/mms/bestpractices/">Sinch MMS channel best practices</see>.
+        ///     If set to true, your message will be rejected if it doesn't conform to the listed
+        ///     recommendations, otherwise no validation will be performed.
+        /// </summary>
+        public bool StrictValidation { get; set; } = false;
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs:31]: Not tagged as required from specs and default value should be managed at server side

File: src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        
+        /// <summary>
+        ///     Contains the parameters that will be used for customizing the message for each recipient.<br /><br />
+        ///     <see href="https://developers.sinch.com/docs/sms/resources/message-info/message-parameterization">
+        ///         Click here to
+        ///         learn more about parameterization.
+        ///     </see>
+        /// </summary>
+        public Dictionary<string, Dictionary<string, string>> Parameters { get; set; }
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs:40]: Not tagged as required from specs

File: src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        /// <summary>
+        ///     URL to the media file
+        /// </summary>
+        public Uri Url { get; set; }
+
+        /// <summary>
+        ///   The message text. Text only media messages will be rejected, please use SMS instead.  
+        /// </summary>
+        public string Message { get; set; }
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/SendMediaBatchRequest.cs:53]: Not tagged as required from specs

File: src/Sinch/SMS/Batches/Send/SendTextBatchRequest.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#endif
+
+        /// <summary>
+        ///     Contains the parameters that will be used for customizing the message for each recipient.<br /><br />
+        ///     <see href="https://developers.sinch.com/docs/sms/resources/message-info/message-parameterization">
+        ///         Click here to
+        ///         learn more about parameterization.
+        ///     </see>
+        /// </summary>
+        public Dictionary<string, Dictionary<string, string>> Parameters { get; set; }
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Send/SendTextBatchRequest.cs:24]: Not tagged as required from specs

File: src/Sinch/SMS/Batches/Update/UpdateBatchBaseRequest.cs
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Update/UpdateBatchBaseRequest.cs:1]: Specs none of fields are required for update (Text & Binary & SMS)

File: src/Sinch/SMS/Batches/Update/UpdateBinaryBatchRequest.cs
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Update/UpdateBinaryBatchRequest.cs:1]: Fields are optional

File: src/Sinch/SMS/Batches/Update/UpdateMediaBatchRequest.cs
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Update/UpdateMediaBatchRequest.cs:1]: Fields are optional

File: src/Sinch/SMS/Batches/Update/UpdateTextBatchRequest.cs
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Update/UpdateTextBatchRequest.cs:1]: Fields are optional

File: src/Sinch/SMS/Batches/Batch.cs [hunk 7]
```diff
@@ -1,8 +1,239 @@
...
+        public override SmsType Type { get; } = SmsType.MtMedia;
+
+        /// <summary>
+        ///     Default: <c>false</c>.<br/><br/>
+        ///     Whether or not you want the media included in your message to be checked against
+        ///     <see href="https://developers.sinch.com/docs/mms/bestpractices/">Sinch MMS channel best practices</see>.
+        ///     If set to true, your message will be rejected if it doesn't conform to the listed
+        ///     recommendations, otherwise no validation will be performed.
+        /// </summary>
+        public bool StrictValidation { get; set; } = false;
```
Comment (JPPortier) [inline src/Sinch/SMS/Batches/Batch.cs:225]: Default values should be managed at server side

File: src/Sinch/SMS/Batches/Send/BatchBase.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        /// </summary>
+        public string From { get; set; }
+
+        public abstract SmsType Type { get; }
+
+        /// <summary>
+        ///     Request delivery report callback.<br/><br/>
+        ///     Note that delivery reports can be fetched from the API regardless of this setting.
+        /// </summary>
+        public DeliveryReport DeliveryReport { get; set; }
```
[the remark this answers is not in the document]
Comment (Dovchik) [reply src/Sinch/SMS/Batches/Send/BatchBase.cs:88]: All of this and below is an optional fields, there is no `Option` monad in c#, and `Nullable<T>` with `?` is for primitives like  `bool`

[the remark this answers is not in the document]
Comment (JPPortier) [reply src/Sinch/SMS/Batches/Send/BatchBase.cs:88]: Ok. Thanks to highlight the language feature 😕

File: src/Sinch/SMS/Batches/Send/BatchBase.cs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        public required List<string> To { get; set; }
+#else
+        public List<string> To { get; set; }
+#endif
+
+        /// <summary>
+        ///     Sender number. Must be valid phone number, short code or alphanumeric.
+        ///     Required if Automatic Default Originator not configured.
+        /// </summary>
+        public string From { get; set; }
```
[the remark this answers is not in the document]
Comment (Dovchik) [reply src/Sinch/SMS/Batches/Send/BatchBase.cs:80]: Can't see where it's required except required by specific condition

File: src/Sinch/SMS/Batches/Send/BatchBase.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dovchik) [reply src/Sinch/SMS/Batches/Send/BatchBase.cs:88]: And I don't see where are they required? Are we using different specs?

File: src/Sinch/SMS/Batches/Send/BatchBase.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (JPPortier) [reply src/Sinch/SMS/Batches/Send/BatchBase.cs:80]: This is the point: not required yes

**Review by JPPortier — APPROVED** [review]

**Pull request merged**